File: InterviewTraining/js/solutions/validateBst.js
/*
Validate Binary Search Tree

Write a function that determines if a binary tree is a valid binary search tree (BST).
A BST is valid if and only if all of its nodes satisfy the BST property.

The BST Property:
- Every node's value must be:
  * STRICTLY GREATER than ALL values in its left subtree
  * STRICTLY LESS than ALL values in its right subtree
- No duplicate values allowed

Example Valid BST:
       10
      /  \
     5    15
    / \   /
   2   7 13

Example Invalid BSTs:

1. Invalid because 15 in left subtree > 10:
       10
      /  \
     15   20

2. Invalid because 5 in right subtree < 10:
       10
      /  \
     3    5

3. Invalid because duplicate value 10:
       10
      /  \
     5    10

Parameters:
- tree: A binary tree node with properties:
  * value: integer
  * left: node or null
  * right: node or null
  * (you may also receive min/max bounds for recursive validation)

Return:
- boolean: true if tree is a valid BST, false otherwise

Hints:
- Think about using min/max bounds for each node
- Remember that ALL nodes in left subtree must be less than parent
- Remember that ALL nodes in right subtree must be greater than parent
*/

function validateBst(tree, minVal = -Infinity, maxVal = Infinity) {
    if (!tree) return true;
    if (tree.value < minVal || tree.value >= maxVal) return false;

    const leftValid = validateBst(tree.left, minVal, tree.value);
    const rightValid = validateBst(tree.right, tree.value, maxVal);

    return leftValid && rightValid;
}

module.exports = validateBst;

File: InterviewTraining/go/solutions/validateBst.go
//go:build problem

package main

import "fmt"

// Solution accepts a tree JSON object and returns whether it's a valid BST.
func Solution(treeObj map[string]interface{}) bool {
    nodesVal, ok := treeObj["nodes"].([]interface{})
    if !ok {
        return false
    }
    m := map[string]map[string]interface{}{}
    var rootId string
    for _, ni := range nodesVal {
        n := ni.(map[string]interface{})
        id := fmt.Sprintf("%v", n["id"])
        m[id] = n
        if rootId == "" {
            rootId = id
        }
    }
    if r, ok := treeObj["rootId"].(string); ok {
        rootId = r
    }
    if r, ok := treeObj["root"].(string); ok {
        rootId = r
    }

    var dfs func(id string, min, max *float64) bool
    dfs = func(id string, min, max *float64) bool {
        if id == "" {
            return true
        }
        n, ok := m[id]
        if !ok {
            return true
        }
        val := n["value"].(float64)
        if min != nil && val <= *min {
            return false
        }
        if max != nil && val >= *max {
            return false
        }
        var leftId, rightId string
        if l, ok := n["left"].(string); ok {
            leftId = l
        }
        if r, ok := n["right"].(string); ok {
            rightId = r
        }
        // left subtree must be < val
        if !dfs(leftId, min, &val) {
            return false
        }
        // right subtree must be > val
        if !dfs(rightId, &val, max) {
            return false
        }
        return true
    }
    return dfs(rootId, nil, nil)
}
